### ISSUE TITLE
Add a flag to determine wether a ReadableStream is disturbed or cancelled

### ISSUE THREAD

**Issue by rumkin** [whatwg/streams#1025]
## Proposal

Implement flags to determine in a synchronous way wether the stream is disturbed and cancelled as boolean properties `ReadableStream#isDisturbed` and `ReadableStream#isCancelled`.

## Rationale

ReadableStream should have a flag which describes if the stream is disturbed or not. Without this flag it's just **impossible** to understand if the first chunk of the stream would be the first chunk of the data. That's why behaviors to check is stream has been read already is implemented in all major browsers ([Chromium](https://github.com/chromium/chromium/blob/40935da55702f890dc4fa0598db880995e965902/third_party/blink/renderer/core/fetch/response.cc#L301), [Safari](https://trac.webkit.org/browser/webkit/branches/safari-609.1.12-branch/Source/WebCore/Modules/fetch/FetchBody.cpp#L67), [Firefox](https://hg.mozilla.org/mozilla-central/file/tip/dom/fetch/Fetch.cpp#l1155)) and is using by FetchAPI (Response and Request constructors).

I know there is an opinion against it. But, as I've already showed, without such flags it's impossible to implement reliable solution. Current design creates asynchronous undefined state and doesn't allow to understand wether the stream is correct constructor argument or not and lead to unpredictable delayed race conditions. As I understand such a solution made due to assumption that stream can be non-exclusively owned and as a result cancelled in any time, what seems incorrect.

## Example

I'm implementing a server which works with Web StreamAP. It requires an ability to check, if the stream contains an acceptable body or not, to prevent throwing parsing errors on valid request and to detect race condition occurred in the process of request handling. Currently I solve this in a hacky way using lowlevel API from the Response object implementation, without this I will emit error each time someone else read the stream without being sure where in the code it has happened.

Solution with Response constructor:

```ts
function isStreamDisturbed(stream: ReadableStream): Boolean {
  try {   
      new Response(stream)
    
      return false
    }
    catch {
      return true
    }
}

const stream = new ReadableStream()
stream.cancel()

isStreamDisturbed(stream) // -> true
```

Of cause this solution is a hack and couldn't be decided as reliable.

**Comment by ricea** [comment]
I am leaning towards adding `isDisturbed` (maybe just as `disturbed`?) but not `isCanceled`. We don't even track internally whether the stream was canceled. From the point of view of the consumer of the stream, it doesn't make a difference whether the stream was canceled or simply read until close.

**Comment by rumkin** [comment]
@ricea Replaced with disturbed and closed.

**Comment by yutakahirano** [comment]
`disturbed` is designed for `Response.bodyUsed` and I'm not sure if we should advertise its use widely. Personally I'm not a fan of the concept...

**Comment by rumkin** [comment]
@yutakahirano It's not only used for `Response.bodyUsed` it's mainly used to understand if response body is valid at all. This is why Response's constructor fails and not just set `bodyUsed` in `true` when the stream has been already disturbed. Without this restriction such things like ServiceWorker's cache wouldn't work correctly.

**Comment by yutakahirano** [comment]
`disturbed` represents whether the stream has ever been read. Validity is (or, at least, can be) a different concept I think.

**Comment by rumkin** [comment]
But in the concept of Request/Response, if the stream is disturbed, then it's not a valid body. So here "disturbed" and "valid" are synonyms.

**Comment by jasnell** [comment]
After having just gone through the implementation for Node.js core, I'd definitely be +1 on a `disturbed` property on `ReadableStream`.

**Comment by benjamingr** [comment]
@ricea @annevk would a spec PR adding a `.disturbed` property on ReadableStream be welcome :)?

**Comment by annevk** [comment]
What would the use case be? As discussed in https://github.com/whatwg/streams/issues/1105 it's rather specific to the Fetch APIs and unclear if we would do that again if starting fresh.

**Comment by benjamingr** [comment]
@annevk honestly Node only needs it for the `fetch` API as well so far https://github.com/nodejs/node/issues/39627 

@ronag is there any other place we need it?

**Comment by ronag** [comment]
> @annevk honestly Node only needs it for the `fetch` API as well so far [nodejs/node#39627](https://github.com/nodejs/node/issues/39627)
> 
> @ronag is there any other place we need it?

So far just for fetch. Though I can see other uses for it. We added `readableDidRead` for node streams and there is `bodyUsed` on body mixin. So I think there is a general need for the concept even if the best practices is a little unclear atm.

**Comment by domenic** [comment]
Still -1 on this; probably it's best if we should close. Fetch is meant to be able to muck with the internals of ReadableStream anyway; keeping them independent is not a design goal.

**Comment by benjamingr** [comment]
Yeah if this is just fetch and nothing else node can just “cheat”

**Comment by domenic** [comment]
Yeah, good reminder to close this issue.
